### PR TITLE
Enable auto dependency inference in spark flavor

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

Enable auto dependency inference in spark flavor.

## How is this patch tested?

Existing tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/_capture_modules.py [hunk 1]
```diff
@@ -84,6 +85,11 @@ def main():
 
     cap_cm = _CaptureImportedModules()
 
+    if flavor == "spark" and is_in_databricks_runtime():
+        from dbruntime.spark_connection import initialize_spark_connection
+
+        initialize_spark_connection()
```
Comment (WeichenXu123) [inline mlflow/utils/_capture_modules.py:91]: if the case not in databricks runtime ?

Comment (harupy) [reply mlflow/utils/_capture_modules.py:91]: @WeichenXu123 

If not, a new spark session is created in the following code:
https://github.com/mlflow/mlflow/blob/c4b8e849a3c718575e106f3499a12bea9b0b821e/mlflow/spark.py#L687-L700

File: mlflow/utils/_capture_modules.py [hunk 2]
```diff
@@ -84,6 +85,11 @@ def main():
 
     cap_cm = _CaptureImportedModules()
 
+    if flavor == "spark" and is_in_databricks_runtime():
+        from dbruntime.spark_connection import initialize_spark_connection
```
Comment (harupy) [inline mlflow/utils/_capture_modules.py:89]: This approach breaks when `initialize_spark_connection` is renamed or moved to a different module.

Comment (harupy) [reply mlflow/utils/_capture_modules.py:89]: Should we make the spark initialization process modifiable via patching or an environment variable?

Comment (WeichenXu123) [reply mlflow/utils/_capture_modules.py:89]: Add a MLR test to prevent it break.

File: mlflow/utils/_capture_modules.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/utils/_capture_modules.py:91]: Shall we add try/catch here and add fallback handling ?
e.g. the case  User upgrade/downgrade the builtin mlflow of databrick runtime and found the `dbruntime.spark_connection.initialize_spark_connection` API does not exist.

Comment (dbczumar) [reply mlflow/utils/_capture_modules.py:91]: Agree with @WeichenXu123 , I think a try/catch is a good idea here.

**Review by dbczumar — APPROVED** [review]
@harupy LGTM once https://github.com/mlflow/mlflow/pull/4759/files#r699873541 is addressed. Great work!

File: mlflow/utils/_capture_modules.py [hunk 3]
```diff
@@ -82,6 +83,15 @@ def main():
...
     sys.path = json.loads(args.sys_path)
 
+    if flavor == mlflow.spark.FLAVOR_NAME and is_in_databricks_runtime():
+        try:
+            # pylint: disable=import-error
+            from dbruntime.spark_connection import initialize_spark_connection
+
+            initialize_spark_connection()
+        except Exception:
+            pass
```
Comment (harupy) [inline mlflow/utils/_capture_modules.py:93]: @WeichenXu123 @dbczumar Added `try-catch`. We could optimistically run the subsequent process after the failure, but `mlflow.spark.load_pyfucn` will also propbably fail If `initialize_spark_connection` fails.

Comment (dbczumar) [reply mlflow/utils/_capture_modules.py:93]: @harupy Makes sense that it will likely fail. Can we raise a clearer error message for debugging purposes?

Comment (harupy) [reply mlflow/utils/_capture_modules.py:93]:
```suggestion
        try:
            # pylint: disable=import-error
            from dbruntime.spark_connection import initialize_spark_connection

            initialize_spark_connection()
        except Exception as e:
            raise Exception("Failed to initialize spark connection") from e
```

Can we just say `Failed to initialize spark connection`?

Comment (dbczumar) [reply mlflow/utils/_capture_modules.py:93]: Sounds good to me!

**Pull request merged**